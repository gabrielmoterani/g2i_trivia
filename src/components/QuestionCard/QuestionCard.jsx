import React, {useState, useEffect} from 'react'
import { View, StyleSheet, Text,TouchableOpacity } from 'react-native';


// Libs Import
import { LinearGradient } from 'expo-linear-gradient';
const Entities = require('html-entities').AllHtmlEntities;

// Static Import 
import COLORS from '../../config/theme';

const QuestionCard = ({gradientColor, question, onSelectAnswer}) => {
    const [selectedAnswer, setSelectedAnswer] = useState('')
    const [isCorrect, setIsCorrect] = useState();
    const correctAnswer = question.correct_answer
    const answers = ['True', 'False'];
    const entities = new Entities();

    useEffect(() => {
        setSelectedAnswer('');
        setIsCorrect(null);
    }, [question])

    const handleSelect = async (answer) => {
        setSelectedAnswer(answer);
        let result = correctAnswer === answer ? true : false;
        setIsCorrect(result)
        onSelectAnswer(result);
    }

    const setAnswerStyle = (answer) => {
        let style = styles.answerBtn;
        if(selectedAnswer === '') return style;
        if(answer === correctAnswer){
            style = {...style, backgroundColor: COLORS.green, borderWidth: 0}
        } else {
            style = {...style, backgroundColor: COLORS.pink, borderWidth: 0}
        }
        return style
    }

    return ( 
        <View style={styles.container}>
            <View style={styles.header(gradientColor)}>
                <Text style={styles.title}>{entities.decode(question.question)}</Text>
            </View>
            <View style={styles.answersContainer}>
                {answers.map(answer => (
                    <TouchableOpacity
                        disabled={selectedAnswer !== ''}
                        onPress={() => {handleSelect(answer)}}
                        key={answer} style={setAnswerStyle(answer)}
                    >
                       <Text style={selectedAnswer !== '' ? {color: COLORS.primary} : ''}>{answer}</Text> 
                    </TouchableOpacity>
                ))}
            </View>
        </View>
     );
}

const styles = StyleSheet.create({
    container: {
      alignItems: 'center',
      backgroundColor: COLORS.primary,
      borderRadius: 10,
      marginTop: 120,
      width: '80%',
      height: 250
    }, 
    header: (gradientColor) => ({
        backgroundColor: gradientColor[0],
        padding: 20,
        borderTopRightRadius: 10,
        borderTopLeftRadius: 10
    }),
    title:{
        color: COLORS.primary
    },
    answersContainer: {
        flex: 1,
        alignSelf: 'stretch',
        justifyContent: 'space-around'
    },
    answerBtn:{
        marginHorizontal: 20,
        alignItems: 'center',
        borderRadius: 20,
        borderColor: COLORS.soft,
        borderWidth: 1,
        paddingVertical: 10
    }
});
 
export default QuestionCard;