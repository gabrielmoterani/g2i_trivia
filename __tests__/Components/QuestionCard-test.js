import React from 'react';
import { shallow } from 'enzyme';
import QuestionCard from '../../src/components/QuestionCard';



initialProps = {
    gradientColor: '#fff',
    question:     
    {
        "category": "Entertainment: Video Games",
        "type": "boolean",
        "difficulty": "hard",
        "question": "In Undertale, having a &quot;Fun Value&quot; set to 56-57 will play the &quot;Wrong Number Song Call&quot;.",
        "correct_answer": "False",
        "incorrect_answers": [
          "True"
        ]
    },
    onSelectAnswer: jest.fn()
}


const setup = (props = initialProps) => shallow(<QuestionCard  {...props} />);


describe('Test QuestionCard Component', () => {
    const wrapper = setup();
  
    it('should render properly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should have two btns', () => {
        const btns = wrapper.find('TouchableOpacity');
        expect(btns.length).toBeGreaterThanOrEqual(2);
    });
    
    it('should call onSelectAnswer', () => {
        const btn = wrapper.find('TouchableOpacity').first();
        btn.simulate( 'press' );

        expect(initialProps.onSelectAnswer).toHaveBeenCalled();
    });

});