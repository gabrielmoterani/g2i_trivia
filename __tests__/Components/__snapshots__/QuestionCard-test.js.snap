// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test QuestionCard Component should render properly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderRadius": 10,
      "height": 250,
      "marginTop": 120,
      "width": "80%",
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#",
        "borderTopLeftRadius": 10,
        "borderTopRightRadius": 10,
        "padding": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
        }
      }
    >
      In Undertale, having a "Fun Value" set to 56-57 will play the "Wrong Number Song Call".
    </Text>
  </View>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "flex": 1,
        "justifyContent": "space-around",
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      disabled={false}
      key="True"
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#333",
          "borderRadius": 20,
          "borderWidth": 1,
          "marginHorizontal": 20,
          "paddingVertical": 10,
        }
      }
    >
      <Text
        style=""
      >
        True
      </Text>
    </TouchableOpacity>
    <TouchableOpacity
      activeOpacity={0.2}
      disabled={false}
      key="False"
      onPress={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "#333",
          "borderRadius": 20,
          "borderWidth": 1,
          "marginHorizontal": 20,
          "paddingVertical": 10,
        }
      }
    >
      <Text
        style=""
      >
        False
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;
